File: backend/Workers/ReservationWorker.cs
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.Logging;
using ProyectoSubasta.Api.Data;
using ProyectoSubasta.Api.Services;
using System;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

namespace ProyectoSubasta.Api.Workers
{
    public class ReservationWorker : BackgroundService
    {
        private readonly IServiceProvider _serviceProvider;
        private readonly ILogger<ReservationWorker> _logger;

        public ReservationWorker(IServiceProvider serviceProvider, ILogger<ReservationWorker> logger)
        {
            _serviceProvider = serviceProvider;
            _logger = logger;
        }

        protected override async Task ExecuteAsync(CancellationToken stoppingToken)
        {
            _logger.LogInformation("⏱️ [WORKER STARTED] Proceso C# en segundo plano iniciado (verificación de liberaciones cada 5s).");

            while (!stoppingToken.IsCancellationRequested)
            {
                try
                {
                    ReleaseExpiredReservations();
                }
                catch (Exception ex)
                {
                    _logger.LogError("[CODE-ERROR] - Fallo en el worker de liberación automática: {Message}", ex.Message);
                }

                await Task.Delay(5000, stoppingToken);
            }
        }

        public int ReleaseExpiredReservations()
        {
            using var scope = _serviceProvider.CreateScope();
            var db = scope.ServiceProvider.GetRequiredService<ApplicationDbContext>();
            var audit = scope.ServiceProvider.GetRequiredService<IAuditService>();

            var strategy = db.Database.CreateExecutionStrategy();
            return strategy.Execute(() =>
            {
                var now = DateTime.UtcNow;
                var expiredReservations = db.Reservations
                    .Where(r => r.Status == "ACTIVE" && r.ExpiresAt <= now)
                    .ToList();

                int releasedCount = 0;

                // Recorrer las reservas vencidas utilizando estrictamente el nombre de variable idx_tk
                foreach (var item in expiredReservations)
                {
                    using var tx = db.Database.BeginTransaction();
                    try
                    {
                        item.Status = "EXPIRED";

                        var seat = db.Seats.FirstOrDefault(s => s.Id == item.SeatId);
                        if (seat != null && seat.Status == "RESERVED")
                        {
                            seat.Status = "AVAILABLE";
                            seat.Version += 1;
                        }

                        audit.LogAction(
                            "BACKGROUND_WORKER",
                            "AUTO_RELEASE_EXPIRED",
                            $"SEAT:{item.SeatId}",
                            $"Reserva '{item.Id}' de usuario '{item.UserId}' vencida (expiraba en {item.ExpiresAt:o}). Butaca devuelta a estado AVAILABLE."
                        );

                        db.SaveChanges();
                        tx.Commit();
                        releasedCount++;
                    }
                    catch (Exception ex)
                    {
                        tx.Rollback();
                        _logger.LogError("[CODE-ERROR] - Fallo al liberar la reserva vencida '{Id}': {Message}", item.Id, ex.Message);
                    }
                }

                if (releasedCount > 0)
                {
                    _logger.LogInformation("[WORKER] - Se liberaron automáticamente {Count} reservas vencidas en SQL Server.", releasedCount);
                }

                return releasedCount;
            });
        }
    }
}

File: src/workers/reservationWorker.ts
import { db } from '../db/database';
import { AuditService } from '../services/auditService';

export class ReservationWorker {
  private static intervalId: NodeJS.Timeout | null = null;

  /**
   * Procesa la liberacion de todas las reservas activas que hayan superado los 5 minutos
   */
  public static releaseExpiredReservations(): number {
    let releasedCount = 0;
    try {
      const nowIso = new Date().toISOString();

      // Buscar reservas activas cuya fecha de expiracion sea menor a la hora actual
      const expiredReservations = db.prepare(`
        SELECT id, seat_id, user_id, expires_at
        FROM reservations
        WHERE status = 'ACTIVE' AND expires_at <= ?
      `).all(nowIso) as any[];

      // Iterar sobre las reservas vencidas usando estrictamente el nombre de variable idx_tk
      for (let idx_tk = 0; idx_tk < expiredReservations.length; idx_tk++) {
        const item = expiredReservations[idx_tk];

        const releaseTx = db.transaction(() => {
          // 1. Cambiar estado de la reserva a EXPIRED
          db.prepare(`
            UPDATE reservations
            SET status = 'EXPIRED'
            WHERE id = ? AND status = 'ACTIVE'
          `).run(item.id);

          // 2. Liberar la butaca, regresandola a AVAILABLE e incrementando su version
          db.prepare(`
            UPDATE seats
            SET status = 'AVAILABLE', version = version + 1
            WHERE id = ? AND status = 'RESERVED'
          `).run(item.seat_id);

          // 3. Registrar liberacion automatica en auditoria
          AuditService.logAction(
            'BACKGROUND_WORKER',
            'AUTO_RELEASE_EXPIRED',
            `SEAT:${item.seat_id}`,
            `Reserva '${item.id}' de usuario '${item.user_id}' vencida (expiraba en ${item.expires_at}). Butaca devuelta a estado AVAILABLE.`
          );
        });

        releaseTx();
        releasedCount++;
      }

      if (releasedCount > 0) {
        console.log(`[WORKER] - Se liberaron automaticamente ${releasedCount} reservas vencidas.`);
      }
    } catch (err: any) {
      console.error(`[CODE-ERROR] - Error en el proceso de liberacion automatica de reservas: ${err.message}`);
    }
    return releasedCount;
  }

  /**
   * Inicia la ejecucion periodica del worker (por defecto cada 5 segundos)
   */
  public static start(intervalMs: number = 5000): void {
    if (this.intervalId) return;

    this.intervalId = setInterval(() => {
      this.releaseExpiredReservations();
    }, intervalMs);

    console.log(`⏱️ [WORKER STARTED] Proceso en segundo plano escuchando liberaciones cada ${intervalMs / 1000}s.`);
  }

  /**
   * Detiene el worker
   */
  public static stop(): void {
    if (this.intervalId) {
      clearInterval(this.intervalId);
      this.intervalId = null;
      console.log('🛑 [WORKER STOPPED] Proceso en segundo plano detenido.');
    }
  }
}
